rust/candid/src/types/subtype.rs

```diff
@@ -0,0 +1,130 @@
+use super::internal::{find_type, Field, Label, Type};
+use crate::parser::typing::TypeEnv;
+use std::collections::{HashMap, HashSet};
+
+/// Check if t1 <: t2
+pub fn subtype(env: &TypeEnv, t1: &Type, t2: &Type) -> bool {
+    use Type::*;
+    if t1 == t2 {
+        return true;
+    }
+    match (t1, t2) {
+        (_, Var(id)) => subtype(env, t1, env.rec_find_type(id).unwrap()),
+        (Var(id), _) => subtype(env, env.rec_find_type(id).unwrap(), t2),
+        (_, Knot(id)) => subtype(env, t1, &find_type(id).unwrap()),
+        (Knot(id), _) => subtype(env, &find_type(id).unwrap(), t2),
+        (_, Reserved) => true,
+        (Empty, _) => true,
+        (Nat, Int) => true,
+        (Vec(ty1), Vec(ty2)) => subtype(env, ty1, ty2),
+        (Null, Opt(_)) => true,
+        (Opt(ty1), Opt(ty2)) if subtype(env, ty1, ty2) => true,
+        (t1, Opt(ty2)) if subtype(env, t1, ty2) && !subtype(env, &Null, ty2) => true,
+        (Opt(ty1), Opt(_ty2)) if !subtype(env, ty1, t2) => {
+            eprintln!("{} <: {} via special opt rule", t1, t2);
+            true
+        }
+        (Opt(ty1), Opt(ty2)) if !subtype(env, ty1, ty2) && !subtype(env, &Null, ty1) => {
+            eprintln!("{} <: {} via special opt rule", t1, t2);
+            true
+        }
+        (Record(fs1), Record(fs2)) => {
+            let fields: HashMap<_, _> = fs1.iter().map(|Field { id, ty }| (id, ty)).collect();
+            for Field { id, ty: ty2 } in fs2.iter() {
+                match fields.get(id) {
+                    Some(ty1) => {
+                        if !subtype(env, ty1, ty2) {
+                            return false;
+                        }
+                    }
+                    None => {
+                        if !subtype(env, &Opt(Box::new(Empty)), ty2) {
+                            return false;
+                        }
+                    }
+                }
+            }
+            true
```

Comment:
Maybe nicer when using `fs2.iter().all(|Field {id, ty : ty2 }| … )`? Or is that just my functional programming fetish at work?

Reply:
Thanks, that's indeed better!